src/components/impagination.js

```diff
@@ -50,156 +51,113 @@ export default class Impagination extends Component {
+      stats: { totalPages: this.props.collection.totalPages }
+    })
+  };
 
-  // set the initial read offset of our dataset
+  // update state on mount to trigger any initial requests and
+  // load any already-resolved records
   componentWillMount() {
-    let { readOffset } = this.props;
-    this.dataset.setReadOffset(readOffset);
+    this.updateDataset();
   }
 
-  // when we recieve an update, loop over the pages of collections so
-  // we can resolve or reject any pending requests
+  // the main trigger for updating the datset state
   componentWillReceiveProps(nextProps) {
-    let { collection } = nextProps;
-
-    // we need to reset the dataset
-    if (collection.key !== this.props.collection.key ||
-        this.hasUnloadedWithinHorizon(nextProps)) {
-      this.promises = {};
-
-      // impagination dataset
-      this.dataset = new Dataset({
-        pageSize: nextProps.pageSize,
-        loadHorizon: nextProps.loadHorizon,
-        stats: { totalPages: collection.totalPages },
-        fetch: this.fetch,
-        observe: this.observe
+    this.updateDataset(nextProps);
+  }
+
+  //  Updates the dataset immutably and triggers the `fetch` property
+  //  when there are unrequested pages.
+  updateDataset(props = this.props) {
+    let { collection, pageSize, loadHorizon, readOffset, fetch } = props;
+    let { dataset } = this.state;
+
+    let isNewCollection = collection.key !== this.props.collection.key;
+    let hasUnloaded = this.getPagesInHorizon().some(({ offset }) => {
+      return collection.getPage(offset + 1).request.hasUnloaded;
+    });
+
+    // we need a brand new dataset state
+    if (isNewCollection || hasUnloaded) {
+      dataset = new State({
+        pageSize,
+        loadHorizon,
+        readOffset,
+        stats: {
+          totalPages: collection.totalPages
+        }
       });
     }
 
-    // if there is a new total page count, update our dataset state
-    if (collection.totalPages !== this.dataset.state.stats.totalPages) {
-      this.dataset.state.stats.totalPages = collection.totalPages;
+    //  update the total page count if necessary
+    if (collection.totalPages !== dataset.stats.totalPages) {
+      dataset.stats.totalPages = collection.totalPages;
     }
 
-    // ensure that we keep all of our promises
-    for (let page of Object.keys(this.promises)) {
-      let { request, records } = collection.getPage(page);
-      let promise = this.promises[page];
+    // update the read offset if necessary
+    if (readOffset !== dataset.readOffset) {
+      dataset = dataset.setReadOffset(readOffset);
+    }
+
+    // make any requests for unrequested pages
+    dataset.unrequested.forEach((p) => {
+      // we use a 1-based page number
+      fetch(p.offset + 1, pageSize);
+    });
+
+    // update newly requested pages
+    dataset = dataset.fetch(dataset.unrequested);
+
+    // resolve or reject all pages within the load horizon
+    this.getPagesInHorizon(dataset).forEach((page) => {
```

Comment:
Do we want to only handle pages in the horizon? or do we want to consider each page with an outstanding request. (isPending)

Also, what about unfetchable pages? I.e. pages outside the load horizon that we can unload.

Reply:
So we don't want to only consider pending requests, because when a single package is updated, the page's request is still resolved and does not change. So we would still have the problem where selecting a package does not get updated in the list.

If we do this to all pages outside of the load horizon, this gets very slow to update records that are outside of the view after you've scrolled down a long list. I think within the horizon is fine because this component is just for the _display_ of data. And we would never need to know the status of dataset records outside of the view. Especially since our records have their own `isResolved` and `isRejected` properties.

As for unfetchable pages, I do think I should add some logic around unfetching pages with errors. But otherwise what would be the benefit of unloading previous records in the list? If we unload them and you end up on the page of an unloaded record, we lose the snapyness of having that data already available in the store.

Reply:
I guess the edge-case I'm thinking of is when we have page requested, and then it leaves the horizon. Will it get stuck in a pending state if it never gets resolved? Or will we resolve it once it moves back into the horizon.

Also, this seems like a candidate for moving upstream. Can we wrap our dataset state to provide a `horizonPages` property so that the interface won't change if we open up an upstream PR?

Reply:
The data in our store will be resolved whenever the request is resolved. The dataset page object will get resolved once it's back in the horizon and the data in our store is also resolved.

Wil do on that `horizonPages` property

Reply:
maybe call it `pagesWithinHorizon` so that it's more easily understandable.